src/ConfigurationClientManager.ts

```diff
@@ -0,0 +1,296 @@
+            }
+        } else if ((connectionStringOrEndpoint instanceof URL || typeof connectionStringOrEndpoint === "string") && instanceOfTokenCredential(credentialOrOptions)) {
+            let endpoint = connectionStringOrEndpoint;
+            // ensure string is a valid URL.
+            if (typeof endpoint === "string") {
+                try {
+                    endpoint = new URL(endpoint);
+                } catch (error) {
+                    if (error.code === "ERR_INVALID_URL") {
+                        throw new Error("Invalid endpoint URL.", { cause: error });
+                    } else {
+                        throw error;
+                    }
+                }
+            }
+
+            const credential = credentialOrOptions as TokenCredential;
+            this.#appConfigOptions = appConfigOptions as AzureAppConfigurationOptions;
+            this.#clientOptions = getClientOptions(this.#appConfigOptions);
+            staticClient = new AppConfigurationClient(connectionStringOrEndpoint.toString(), credential, this.#clientOptions);
+            this.endpoint = endpoint.toString();
+            this.#credential = credential;
+        } else {
+            throw new Error("A connection string or an endpoint with credential must be specified to create a client.");
+        }
+
+        this.#staticClients = [new ConfigurationClientWrapper(this.endpoint, staticClient)];
+        this.#validDomain = getValidDomain(this.endpoint);
+    }
+
+    async init() {
+        if (this.#appConfigOptions?.replicaDiscoveryEnabled === false || isBrowser() || isWebWorker()) {
+            this.isFailoverable = false;
+            return;
+        }
+
+        try {
+            this.dns = await import("dns/promises");
+        }catch (error) {
+            this.isFailoverable = false;
+            console.warn("Failed to load the dns module:", error.message);
+            return;
+        }
+
+        this.isFailoverable = true;
+    }
+
+    async getClients() : Promise<ConfigurationClientWrapper[]> {
+        if (!this.isFailoverable) {
+            return this.#staticClients;
+        }
+
+        const currentTime = Date.now();
+        if (this.#isFallbackClientDiscoveryDue(currentTime)) {
+            this.#lastFallbackClientRefreshAttempt = currentTime;
+            const host = new URL(this.endpoint).hostname;
+            await this.#discoverFallbackClients(host);
+        }
+
+        // Filter static clients whose backoff time has ended
+        let availableClients = this.#staticClients.filter(client => client.backoffEndTime <= currentTime);
+        // If there are dynamic clients, filter and concatenate them
+        if (this.#dynamicClients && this.#dynamicClients.length > 0) {
+            availableClients = availableClients.concat(
+                this.#dynamicClients
+                    .filter(client => client.backoffEndTime <= currentTime));
```

Comment:
For the code path where `isFallbackClientDiscoveryDue` is true and`discoverFallbackClients` is called, `this.#dynamicClients` will be all brand-new clients.

The `filter` call can be saved.

``` javascript
 
let availableClients = this.#staticClients.filter(client => client.backoffEndTime <= currentTime);
if (this.#isFallbackClientDiscoveryDue()){
    await this.#discoverFallbackClients();
    return availableClients.concat(this.#dynamicClients)
}
return availableClients.concat(this.#dynamicClients.filter(client => client.backoffEndTime <= currentTime));
```

Reply:
Related to this [conversation](https://github.com/Azure/AppConfiguration-JavaScriptProvider/pull/98/files#r1838657409)
This can be improved even more if you implement `isFallbackClientDiscoveryDue` inline here. 
Then, we can only do something like this:

``` javascript
const availableReplicas = this.#dynamicClients.filter(client => client.backoffEndTime <= currentTime);
if (dateTime >= this.#lastFallbackClientRefreshAttempt + MINIMAL_CLIENT_REFRESH_INTERVAL
            && (availableReplicas.length == 0 || Date.now() >= this.#lastFallbackClientRefreshTime + FALLBACK_CLIENT_REFRESH_EXPIRE_INTERVAL)) {
    await this.#discoverFallbackClients();
    return availableClients.concat(this.#dynamicClients)
}
return availableClients.concat(availableReplicas);
```